File: sdk/javascript/src/subscriptions.ts
import type { SorobanPulseClient } from "./client";
import type { EventHandler, SorobanEvent } from "./types";

export interface EventSubscriptionOptions {
  contractId?: string;
  eventTypes?: string[];
  maxReconnects?: number;
}

/**
 * Consumes the SorobanPulse Server-Sent Events (SSE) stream, dispatching
 * parsed events to registered handlers, with automatic reconnection using
 * exponential backoff.
 */
export class EventSubscription {
  private handlers: EventHandler[] = [];
  private stopped = false;

  constructor(
    private readonly client: SorobanPulseClient,
    private readonly options: EventSubscriptionOptions = {},
  ) {}

  onEvent(handler: EventHandler): void {
    this.handlers.push(handler);
  }

  stop(): void {
    this.stopped = true;
  }

  private streamUrl(): string {
    const url = new URL(`${this.client.baseUrl}/events/stream`);
    if (this.options.contractId) url.searchParams.set("contract_id", this.options.contractId);
    for (const eventType of this.options.eventTypes ?? []) {
      url.searchParams.append("event_type", eventType);
    }
    return url.toString();
  }

  private dispatch(raw: string): void {
    const event = JSON.parse(raw) as SorobanEvent;
    for (const handler of this.handlers) handler(event);
  }

  /** Blocking async loop; call without awaiting to run in the background. */
  async run(): Promise<void> {
    const maxReconnects = this.options.maxReconnects ?? 10;
    let attempts = 0;

    while (!this.stopped && attempts < maxReconnects) {
      try {
        const response = await fetch(this.streamUrl(), {
          headers: { Authorization: `Bearer ${(this.client as unknown as { apiKey: string }).apiKey ?? ""}` },
        });
        if (!response.body) throw new Error("no response body for SSE stream");

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = "";

        for (;;) {
          if (this.stopped) return;
          const { done, value } = await reader.read();
          if (done) break;

          buffer += decoder.decode(value, { stream: true });
          const lines = buffer.split("\n");
          buffer = lines.pop() ?? "";

          for (const line of lines) {
            if (line.startsWith("data:")) {
              this.dispatch(line.slice("data:".length).trim());
            }
          }
        }
        attempts = 0;
      } catch {
        attempts += 1;
        const backoffMs = Math.min(2 ** attempts * 1000, 30_000);
        await new Promise((resolve) => setTimeout(resolve, backoffMs));
      }
    }
  }
}

File: sdk/python/soroban_pulse/subscriptions.py
"""Event subscription helpers (SSE streaming)."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Callable, Iterator, Optional
from urllib import request as urllib_request

from .client import SorobanPulseClient


@dataclass
class SorobanEvent:
    id: str
    contract_id: str
    event_type: str
    ledger: int
    data: dict


class EventSubscription:
    """Consumes SorobanPulse's Server-Sent Events (SSE) stream and dispatches
    parsed events to a callback, with automatic reconnection.

    Example:
        sub = EventSubscription(client, contract_id="C...")
        sub.on_event(lambda evt: print(evt.event_type, evt.data))
        sub.run()
    """

    def __init__(self, client: SorobanPulseClient, contract_id: Optional[str] = None,
                 event_types: Optional[list] = None) -> None:
        self.client = client
        self.contract_id = contract_id
        self.event_types = event_types or []
        self._handler: Optional[Callable[[SorobanEvent], None]] = None
        self._stopped = False

    def on_event(self, handler: Callable[[SorobanEvent], None]) -> None:
        self._handler = handler

    def stop(self) -> None:
        self._stopped = True

    def _stream_url(self) -> str:
        params = []
        if self.contract_id:
            params.append(f"contract_id={self.contract_id}")
        for et in self.event_types:
            params.append(f"event_type={et}")
        query = "&".join(params)
        suffix = f"?{query}" if query else ""
        return f"{self.client.base_url}/events/stream{suffix}"

    def _iter_sse_lines(self) -> Iterator[str]:
        req = urllib_request.Request(self._stream_url(), headers=self.client._headers())
        with urllib_request.urlopen(req) as resp:
            for raw_line in resp:
                yield raw_line.decode("utf-8").rstrip("\n")

    def run(self, max_reconnects: int = 10) -> None:
        """Blocking loop that reconnects with exponential backoff on
        transient network failures, up to `max_reconnects` attempts."""
        attempts = 0
        while not self._stopped and attempts < max_reconnects:
            try:
                buffer = ""
                for line in self._iter_sse_lines():
                    if self._stopped:
                        return
                    if line.startswith("data:"):
                        buffer = line[len("data:"):].strip()
                    elif line == "" and buffer:
                        self._dispatch(buffer)
                        buffer = ""
                attempts = 0
            except Exception:
                attempts += 1
                backoff = min(2 ** attempts, 30)
                import time

                time.sleep(backoff)

    def _dispatch(self, raw_json: str) -> None:
        if not self._handler:
            return
        payload = json.loads(raw_json)
        event = SorobanEvent(
            id=payload.get("id", ""),
            contract_id=payload.get("contract_id", ""),
            event_type=payload.get("event_type", ""),
            ledger=payload.get("ledger", 0),
            data=payload.get("data", {}),
        )
        self._handler(event)
